notebooks/langchain.ipynb

```diff
@@ -0,0 +1,218 @@
+      "Time to load the model weights: 23.63 seconds.\n",
+      "You are using a CUDA device ('NVIDIA A100-SXM4-40GB') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision\n"
+     ]
+    }
+   ],
+   "source": [
+    "model, tokenizer, fabric = build_llm(checkpoint_dir=checkpoint_dir, devices=devices, quantize=quantize)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will create a (CustomLLM)[https://python.langchain.com/docs/modules/model_io/models/llms/how_to/custom_llm], which is a callable class and it will be responsible for interacting with our LLM."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class LitGPTLLM(LLM):\n",
+    "    model: Any\n",
+    "    tokenizer: Tokenizer\n",
+    "\n",
+    "    @property\n",
+    "    def _llm_type(self) -> str:\n",
+    "        return \"lit-gpt\"\n",
+    "\n",
+    "    def _call(\n",
+    "        self,\n",
+    "        prompt: str,\n",
+    "        stop: Optional[List[str]] = None,\n",
+    "        run_manager: Optional[CallbackManagerForLLMRun] = None,\n",
+    "    ) -> str:\n",
+    "        if stop is not None:\n",
+    "            raise ValueError(\"stop kwargs are not permitted.\")\n",
+    "        \n",
+    "        encoded = self.tokenizer.encode(prompt, device=self.model.device)\n",
+    "        prompt_length = encoded.size(0)\n",
+    "        max_returned_tokens = prompt_length + max_new_tokens\n",
+    "        assert max_returned_tokens <= self.model.config.block_size, (\n",
+    "            max_returned_tokens,\n",
+    "            self.model.config.block_size,\n",
+    "        )  # maximum rope cache length\n",
+    "        y = generate(\n",
+    "            self.model,\n",
+    "            encoded,\n",
+    "            max_returned_tokens,\n",
+    "            max_seq_length=max_returned_tokens,\n",
+    "            temperature=temperature,\n",
+    "            top_k=top_k,\n",
+    "        ) \n",
+    "        return self.tokenizer.decode(y)\n",
+    "\n",
+    "    @property\n",
+    "    def _identifying_params(self) -> Mapping[str, Any]:\n",
+    "        \"\"\"Get the identifying parameters.\"\"\"\n",
+    "        return {\"name\": self.model.config.name}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "llm = LitGPTLLM(model=model, tokenizer=tokenizer)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 32,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "hello,\n",
```

Comment:
trying a longer prompt leads me to an error for this notebook - `RuntimeError: The size of tensor a (51) must match the size of tensor b (55) at non-singleton dimension 3`


```
File [~/llms/lit-gpt/lit_gpt/generate/base.py:127](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/llms/lit-gpt/lit_gpt/generate/base.py:127), in generate(model, idx, max_returned_tokens, max_seq_length, temperature, top_k, eos_id)
    124 x = idx.index_select(0, input_pos).view(1, -1)
    126 # forward
--> 127 logits = model(x, max_seq_length, input_pos)
    128 logits = logits[0, -1] [/](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/) temperature
    130 # optionally crop the logits to only the top k options

File [~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1502](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1502), in Module._wrapped_call_impl(self, *args, **kwargs)
   1500     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1501 else:
-> 1502     return self._call_impl(*args, **kwargs)

File [~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1511](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1511), in Module._call_impl(self, *args, **kwargs)
   1506 # If we don't have any hooks, we want to skip the rest of the logic in
   1507 # this function, and just call forward.
   1508 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1509         or _global_backward_pre_hooks or _global_backward_hooks
   1510         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1511     return forward_call(*args, **kwargs)
   1512 # Do not call functions when jit is used
   1513 full_backward_hooks, non_full_backward_hooks = [], []

File [~/miniconda3/envs/parrot/lib/python3.11/site-packages/lightning/fabric/wrappers.py:116](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/miniconda3/envs/parrot/lib/python3.11/site-packages/lightning/fabric/wrappers.py:116), in _FabricModule.forward(self, *args, **kwargs)
    113 args, kwargs = self._precision.convert_input((args, kwargs))
    115 with self._precision.forward_context():
--> 116     output = self._forward_module(*args, **kwargs)
    118 output = self._precision.convert_output(output)
    119 return output

File [~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1502](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1502), in Module._wrapped_call_impl(self, *args, **kwargs)
   1500     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1501 else:
-> 1502     return self._call_impl(*args, **kwargs)

File [~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1511](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1511), in Module._call_impl(self, *args, **kwargs)
   1506 # If we don't have any hooks, we want to skip the rest of the logic in
   1507 # this function, and just call forward.
   1508 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1509         or _global_backward_pre_hooks or _global_backward_hooks
   1510         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1511     return forward_call(*args, **kwargs)
   1512 # Do not call functions when jit is used
   1513 full_backward_hooks, non_full_backward_hooks = [], []

File [~/llms/lit-gpt/lit_gpt/model.py:104](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/llms/lit-gpt/lit_gpt/model.py:104), in GPT.forward(self, idx, max_seq_length, input_pos)
    102     self.kv_caches = self.kv_caches or self.build_kv_caches(x, max_seq_length, cos.size(-1))
    103     for i, block in enumerate(self.transformer.h):
--> 104         x, self.kv_caches[i] = block(x, (cos, sin), max_seq_length, mask, input_pos, self.kv_caches[i])
    106 x = self.transformer.ln_f(x)
    108 logits = self.lm_head(x)  # (b, t, vocab_size)

File [~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1502](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1502), in Module._wrapped_call_impl(self, *args, **kwargs)
   1500     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1501 else:
-> 1502     return self._call_impl(*args, **kwargs)

File [~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1511](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1511), in Module._call_impl(self, *args, **kwargs)
   1506 # If we don't have any hooks, we want to skip the rest of the logic in
   1507 # this function, and just call forward.
   1508 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1509         or _global_backward_pre_hooks or _global_backward_hooks
   1510         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1511     return forward_call(*args, **kwargs)
   1512 # Do not call functions when jit is used
   1513 full_backward_hooks, non_full_backward_hooks = [], []

File [~/llms/lit-gpt/lit_gpt/model.py:167](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/llms/lit-gpt/lit_gpt/model.py:167), in Block.forward(self, x, rope, max_seq_length, mask, input_pos, kv_cache)
    157 def forward(
    158     self,
    159     x: torch.Tensor,
   (...)
    164     kv_cache: Optional[KVCache] = None,
    165 ) -> Tuple[torch.Tensor, Optional[KVCache]]:
    166     n_1 = self.norm_1(x)
--> 167     h, new_kv_cache = self.attn(n_1, rope, max_seq_length, mask, input_pos, kv_cache)
    168     if self.config.parallel_residual:
    169         n_2 = n_1 if self.config.shared_attention_norm else self.norm_2(x)

File [~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1502](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1502), in Module._wrapped_call_impl(self, *args, **kwargs)
   1500     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1501 else:
-> 1502     return self._call_impl(*args, **kwargs)

File [~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1511](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/miniconda3/envs/parrot/lib/python3.11/site-packages/torch/nn/modules/module.py:1511), in Module._call_impl(self, *args, **kwargs)
   1506 # If we don't have any hooks, we want to skip the rest of the logic in
   1507 # this function, and just call forward.
   1508 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1509         or _global_backward_pre_hooks or _global_backward_hooks
   1510         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1511     return forward_call(*args, **kwargs)
   1512 # Do not call functions when jit is used
   1513 full_backward_hooks, non_full_backward_hooks = [], []

File [~/llms/lit-gpt/lit_gpt/model.py:247](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/home/aniket/llms/lit-gpt/notebooks/~/llms/lit-gpt/lit_gpt/model.py:247), in CausalSelfAttention.forward(self, x, rope, max_seq_length, mask, input_pos, kv_cache)
    244     kv_cache = k, v
    246 # efficient attention using Flash Attention CUDA kernels
--> 247 y = torch.nn.functional.scaled_dot_product_attention(
    248     q, k, v, attn_mask=mask, dropout_p=0.0, scale=1.0 [/](https://vscode-remote+ssh-002dremote-002b10-002e10-002e10-002e25.vscode-resource.vscode-cdn.net/) math.sqrt(self.config.head_size), is_causal=mask is None
    249 )
    251 y = y.transpose(1, 2).contiguous().view(B, T, C)  # re-assemble all head outputs side by side
    253 # output projection

RuntimeError: The size of tensor a (51) must match the size of tensor b (55) at non-singleton dimension 3
```